### PR TITLE
expression: implement vectorized evaluation for `builtinCastDurationAsRealSig`

### PR DESCRIPTION
**Pull request by pingyu**
PCP #12105 

### What problem does this PR solve? 
Implement vectorized evaluation for builtinCastDurationAsRealSig at #12105 

### What is changed and how it works?
It gets faster about 1.2%.
`Duration.ToNumber()` is modified, just as what is done in `CheckFsp()` from `fsp.go`. Otherwise `builtinCastDurationAsRealSig.evalReal()` in `builtin_cast.go` will raise a `runtime error: slice bounds out of range [:-1]` in unittest of `builtinCastDurationAsRealSig.vecEvalReal()`.

### Check List 

Tests 

 - Unit test
```
> GO111MODULE=on go test -check.f TestVectorizedBuiltinCastFunc
PASS: builtin_cast_vec_test.go:130: testEvaluatorSuite.TestVectorizedBuiltinCastFunc	0.134s
OK: 1 passed
PASS
ok  	github.com/pingcap/tidb/expression	0.158s
```

 - Benchmark
```
> go test -v -benchmem -bench=BenchmarkVectorizedBuiltinCastFunc -run=BenchmarkVectorizedBuiltinCastFunc -args builtinCastDurationAsRealSig
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinCastFunc/builtinCastDurationAsRealSig-VecBuiltinFunc-12         	    4402	    255587 ns/op	   50713 B/op	    3669 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastDurationAsRealSig-NonVecBuiltinFunc-12      	    4347	    277312 ns/op	   50712 B/op	    3669 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	2.430s
```

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -978,7 +978,7 @@ func (d Duration) ToNumber() *MyDecimal {
 		signStr = "-"
 	}
 
-	if d.Fsp == 0 {
```
Comment (wshwsh12) [inline types/time.go]: I don't think we should modify this function. We can use checkfsp() to modify fsp when it is UnspecifiedFsp.

Comment (pingyu) [reply types/time.go]: Comment addressed, PTAL, Thanks. @wshwsh12

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -875,11 +875,42 @@ func (b *builtinCastJSONAsStringSig) vecEvalString(input *chunk.Chunk, result *c
...
+	if err = b.args[0].VecEvalDuration(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	result.ResizeFloat64(n, false)
+	result.MergeNulls(buf)
+	f64s := result.Float64s()
+
+	var duration types.Duration
+	fsp := int8(b.tp.Decimal)
```
Comment (qw4990) [inline expression/builtin_cast_vec.go]: Should it be `b.args[0].GetType().Decimal`?

Comment (pingyu) [reply expression/builtin_cast_vec.go]: Oh...I'm sorry...

Comment addressed, PTAL, Thanks. @qw4990

**Pull request merged**